spikeinterface/core/waveform_extractor.py

```diff
@@ -307,6 +307,25 @@ def get_all_templates(self, unit_ids=None, mode='median', quantile_value=0.5):
             templates[i, :, :] = self.get_template(unit_id, mode=mode, quantile_value=quantile_value)
         return templates
 
+    def get_template_segment(self, unit_id, segment_index, quantile_value=None, mode='median'):
+        assert mode in ('median', 'average', 'std', 'quantile')
+        assert unit_id in self.sorting.unit_ids
+
+        def waveforms_for_segment():
```

Comment:
I would just move this folder outside (and maybe make it private `_waveforms_for_segment(self, index_ar)`). The rest looks good to me!

Reply:
okay, have moved it in https://github.com/SpikeInterface/spikeinterface/pull/258/commits/5b7bf6b026e33d254390720e56db9d24a86ec0f6 and made a separate method, which I have not enforced as internal since I think it could potentially be a useful method for an edge case user/writing all waveforms to nwb.